gax/src/main/java/com/google/api/gax/rpc/ServerStreamIterator.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2017, Google LLC All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.rpc;
+
+import com.google.common.base.Throwables;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+/**
+ * Internal implementation of a blocking Iterator, which will coordinate with the
+ * QueuingResponseObserver to fetch new items from upstream. The Iterator expects the observer to
+ * request the first item, afterwards, new items will be requested when the current ones are
+ * consumed by next().
+ *
+ * <p>Package-private for internal use.
+ *
+ * @param <V> The type of items to be Iterated over.
+ */
+final class ServerStreamIterator<V> implements Iterator<V> {
+  private final QueuingResponseObserver<V> observer;
+  private Object last;
+
+  ServerStreamIterator(QueuingResponseObserver<V> observer) {
+    this.observer = observer;
+  }
+
+  /**
+   * Checks if the next call to {@code hasNext()} (and {@code next()}) will block.
+   *
+   * @return true if the next call is guaranteed to be nonblocking.
+   */
+  boolean isReady() {
+    return last != null || observer.isReady();
+  }
+
+  /**
+   * Consumes the next response and asynchronously request the next response from the observer.
+   *
+   * @return The next response.
+   * @throws NoSuchElementException If the stream has been consumed or cancelled.
+   */
+  @Override
+  public V next() {
```

Comment:
I think you can use https://google.github.io/guava/releases/15.0/api/docs/com/google/common/collect/AbstractIterator.html ?

Reply:
Thanks for pointing it out, I never knew it existed, but I don't think it's appropriate for this usage pattern. The queue idiom is very useful in serializing events and adapting it to the AbstractIterator will create awkwardness in peeking if the next element of the queue is nondata.

Reply:
We couldn't introduce AbstractIterator as-is because it would put Guava on the surface.

Reply:
Oh right, duh, sorry.